.circleci/config.yml

```diff
@@ -93,53 +117,195 @@ commands:
           command: ./scripts/changed-files.sh
 
 jobs:
-  build:
-    executor: build-executor
-    parallelism: 5
+  prefetch-crates:
+    executor: test-executor
+    description: Prefetch cargo crates for subsequent jobs.
     steps:
       - build_setup
       - run:
           name: Git Hooks and Checks
           command: ./scripts/git-checks.sh
       - run:
-          name: Linting
+          name: Fetch workspace dependencies over network
+          command: cargo fetch
+      - save_cargo_package_cache
+      - persist_to_workspace:
+          root: /home/circleci/project
+          paths:
+            - git
+            - registry
+            - .package-cache
+  lint:
+    executor: test-executor
+    description: Run Rust linting tools.
+    steps:
+      - build_setup
+      - run:
+          name: Git Hooks and Checks
+          command: ./scripts/git-checks.sh
+      - attach_workspace:
+          at: /home/circleci/project
+      - restore_cargo_package_cache
+      - run:
+          name: cargo lint
+          command: cargo x lint
+      - run:
+          name: cargo clippy
+          command: cargo xclippy --workspace --all-targets
+      - run:
+          name: cargo fmt
+          command: cargo xfmt --check
+      - run:
+          name: cargo guppy
           command: |
-            [[ $CIRCLE_NODE_INDEX =~ [1234] ]] || cargo x lint
-            [[ $CIRCLE_NODE_INDEX =~ [1234] ]] || cargo xclippy --workspace --all-targets
-            [[ $CIRCLE_NODE_INDEX =~ [1234] ]] || cargo xfmt --check
-            [[ $CIRCLE_NODE_INDEX =~ [1234] ]] || cargo install cargo-guppy --git http://github.com/calibra/cargo-guppy --rev 8b2bc45c0cd6323a7a2b8170ddad6d2a5b79047b
-            [[ $CIRCLE_NODE_INDEX =~ [1234] ]] || [[ -z $(cargo guppy dups --target x86_64-unknown-linux-gnu --kind directthirdparty) ]]
+            cargo install cargo-guppy \
+              --git http://github.com/calibra/cargo-guppy \
+              --rev 8b2bc45c0cd6323a7a2b8170ddad6d2a5b79047b
+            [[ -z $(cargo guppy dups --target x86_64-unknown-linux-gnu \
+              --kind directthirdparty) ]]
+  build-dev:
+    executor: build-executor
+    description: Development Build
+    steps:
+      - build_setup
+      - attach_workspace:
+          at: /home/circleci/project
+      - restore_cargo_package_cache
       - run:
-          name: Build Release
+          name: Biuld default
           command: |
-            [[ $CIRCLE_NODE_INDEX =~ [0234] ]] || RUST_BACKTRACE=1 cargo build -j 16 --release
+            RUST_BACKTRACE=1 cargo build -j 16
```

Comment:
How many cores do we have :O

Reply:
We can configure the job to use up to 20 (https://circleci.com/docs/2.0/configuration-reference/#docker-executor). 

I keep the same 16 CPU option for rust builds in this config so that we can measure how much parallelization buys us.  We can do the cost-benefit analysis and bump it to 20 in the future.